module/input/PS3Controller/src/PS3Controller.cpp

```diff
@@ -0,0 +1,244 @@
+        START          = 9,
+        PS             = 10,
+        JOYSTICK_LEFT  = 11,
+        JOYSTICK_RIGHT = 12,
+        DPAD_UP        = 13,
+        DPAD_DOWN      = 14,
+        DPAD_LEFT      = 15,
+        DPAD_RIGHT     = 16
+    };
+    enum class Axis : uint8_t {
+        JOYSTICK_LEFT_HORIZONTAL = 0,
+        JOYSTICK_LEFT_VERTICAL,
+        L2,
+        JOYSTICK_RIGHT_HORIZONTAL,
+        JOYSTICK_RIGHT_VERTICAL,
+        R2,
+        ACCELEROMETER_X = 23,
+        ACCELEROMETER_Y,
+        ACCELEROMETER_Z
+    };
+
+    using extension::Configuration;
+
+    using message::input::PS3Controller::Accelerometer;
+    using message::input::PS3Controller::CircleButton;
+    using message::input::PS3Controller::CrossButton;
+    using message::input::PS3Controller::DPadButton;
+    using message::input::PS3Controller::L1Button;
+    using message::input::PS3Controller::L2Trigger;
+    using message::input::PS3Controller::LeftJoystick;
+    using message::input::PS3Controller::LeftJoystickButton;
+    using message::input::PS3Controller::PSButton;
+    using message::input::PS3Controller::R1Button;
+    using message::input::PS3Controller::R2Trigger;
+    using message::input::PS3Controller::RightJoystick;
+    using message::input::PS3Controller::RightJoystickButton;
+    using message::input::PS3Controller::SelectButton;
+    using message::input::PS3Controller::SquareButton;
+    using message::input::PS3Controller::StartButton;
+    using message::input::PS3Controller::TriangleButton;
+
+    PS3Controller::PS3Controller(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
+
+        on<Configuration>("PS3Controller.yaml").then([this](const Configuration& config) {
+            // Only reconnect if we are changing the path to the device
+            if ((controller_path.compare(config["controller_path"]) != 0)
+                || (accelerometer_path.compare(config["accelerometer_path"]) != 0)) {
+                controller_path    = config["controller_path"];
+                accelerometer_path = config["accelerometer_path"];
+                connect();
+            }
+        });
+
+        on<Shutdown>().then([this] { disconnect(); });
+
+        // Details on Linux Joystick API
+        // https://www.kernel.org/doc/Documentation/input/joystick-api.txt
+        // Trigger when the joystick has an data to read
+        on<IO>(controller_fd, IO::READ).then([this] {
+            // JoystickEvent has
+            // Timestamp: uint32_t
+            // Value:     int16_t
+            // Type:      uint8_t
+            // Number:    uint8_t
+            constexpr size_t JOYSTICK_EVENT_SIZE =
+                sizeof(uint32_t) + sizeof(int16_t) + sizeof(uint8_t) + sizeof(uint8_t);
+
+            // Try to read a JoystickEvent worth of data
+            std::array<uint8_t, JOYSTICK_EVENT_SIZE> buffer;
+            auto num_bytes = read(controller_fd, buffer.data(), buffer.size());
```

Comment:
This only reads a single packet, you should keep reading these until there are no more. The way it works (you’re reading from a kernel driver not a serial device) you’ll always get exactly this size packet. This also means you can just read into a pragma packed struct to avoid the later reinterpret casts